docs: make structural directives guide generic (#44895)
angular/angular: aio/content/guide/structural-directives.md

# Structural directives

> This guides talks about structural directives and provides conceptual information on how such directives work, how Angular interprets their shorthand syntax, and how to add template guard properties to catch template type errors.

Reviewer: ```suggestion
This guide is about structural directives and provides conceptual information on how such directives work, how Angular interprets their shorthand syntax, and how to add template guard properties to catch template type errors.
```

Author: :facepalm::+1:

<div class="alert is-helpful">

For the example application that this page describes, see the <live-example></live-example>.

</div>

## What they are

Structural directives are directives which change the DOM layout by adding and removing DOM element.

> Angular provides a set of built-in structural directives (such as `NgIf`, `NgFor`, `NgSwitch` and others) which are commonly used in all Angular projects, for more information see [Built-in directives](guide/built-in-directives).

Reviewer: ```suggestion
Angular provides a set of built-in structural directives (such as `NgIf`, `NgFor`, `NgSwitch` and others) which are commonly used in all Angular projects. For more information see [Built-in directives](guide/built-in-directives).
```

{@a shorthand}
{@a asterisk}

## Structural directive shorthand

> When structural directives are applied they generally are prefixed by an asterisk, `*`,  such as `*ngIf`, such convention is shorthand that Angular interprets and converts into a longer form.

Reviewer: ```suggestion
When structural directives are applied they generally are prefixed by an asterisk, `*`,  such as `*ngIf`. This convention is shorthand that Angular interprets and converts into a longer form.
```

> Substantially Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that surrounds the host element and its descendants.

Reviewer: ```suggestion
Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that surrounds the host element and its descendants.
```

> Let's for example take the following code which uses an `*ngIf` to displays the hero's name if `hero` exists:

Reviewer: ```suggestion
For example, let's take the following code which uses an `*ngIf` to displays the hero's name if `hero` exists:
```

# Structural directives

This guides talks about structural directives and provides conceptual information on how such directives work, how Angular interprets their shorthand syntax, and how to add template guard properties to catch template type errors.

<div class="alert is-helpful">

For the example application that this page describes, see the <live-example></live-example>.

</div>

> ## What they are

Author: may not be the best section name ever.... :sweat:

Reviewer 2: You could eliminate the need for this header by moving lines 13-15 to above the example. 
Because that's the obvious place to see a description of the subject you're about to discuss further in the topic, it doesn't need a heading to introduce it.

Structural directives are directives which change the DOM layout by adding and removing DOM element.

> Angular offers a set of built-in structural directives which are widely used in all Angular projects, for more information see [Built-in directives](guide/built-in-directives).

Reviewer 3: ```suggestion
Angular provides a set of built-in structural directives (such as `NgIf`, `NgFor`, `NgSwitch` and others) which are commonly used in all Angular projects, for more information see [Built-in directives](guide/built-in-directives).
```

{@a shorthand}
{@a asterisk}

## Structural directive shorthand

When structural directives are applied they generally are prefixed by an asterisk, `*`,  such as `*ngIf`, such convention is shorthand that Angular interprets and converts into a longer form.
Substantially Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that surrounds the host element and its descendants.

Let's for example take the following code which uses an `*ngIf` to displays the hero's name if `hero` exists:

<code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (asterisk)" region="asterisk"></code-example>

> What Angular does in this case is creating an `<ng-template>` element and moving the `*ngIf` directive onto it where it becomes a property binding in square brackets, `[ngIf]`. The rest of the `<div>`, including its class attribute, is then moved inside the `<ng-template>`:

Reviewer 3: ```suggestion
Angular creates an `<ng-template>` element and applies the `*ngIf` directive onto it where it becomes a property binding in square brackets, `[ngIf]`. The rest of the `<div>`, including its class attribute, is then moved inside the `<ng-template>`:
```

<code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (ngif-template)" region="ngif-template"></code-example>

Note that Angular does not actually create a real `<ng-template>` element, but instead only renders the `<div>` and adds a comment node placeholder to the DOM.

```html

> <!--bindings={
>   "ng-reflect-ng-if": "[object Object]"
> }-->
> <div _ngcontent-c0>Mr. Nice</div>
> 

Reviewer 3: I think we can omit HTML comments, they are not load bearing:
```suggestion
<div _ngcontent-c0>Mr. Nice</div>
```

Author: I think the comment was there just to show what's explained in the line above
> ... adds a comment node placeholder ...

so I actually don't thing that it is too bad, do you find this not a useful info for the user? should we also remove the text part talking about the comment node?

Reviewer 3: Yeah, I think we can remove the reference to the comment node. It's an implementation detail, so we should not highlight it in a guide.

The `let` keyword declares a template input variable that you can reference within the template.
The input variables in this example are `hero`, `i`, and `odd`.
The parser translates `let hero`, `let i`, and `let odd` into variables named `let-hero`, `let-i`, and `let-odd`.
The `let-i` and `let-odd` variables become `let i=index` and `let odd=odd`.
Angular sets `i` and `odd` to the current value of the context's `index` and `odd` properties.

The parser applies PascalCase to all directives and prefixes them with the directive's attribute name, such as ngFor.
For example, the `ngFor` input properties, `of` and `trackBy`, map to `ngForOf` and `ngForTrackBy`.

As the `NgFor` directive loops through the list, it sets and resets properties of its own context object.
These properties can include, but aren't limited to, `index`, `odd`, and a special property
named `$implicit`.

Angular sets `let-hero` to the value of the context's `$implicit` property, which `NgFor` has initialized with the hero for the current iteration.

For more information, see the [NgFor API](api/common/NgForOf "API: NgFor") and [NgForOf API](api/common/NgForOf) documentation.

<div class="alert is-helpful">

>   Note that Angular's `<ng-template>` element defines a template that doesn't render anything by default, if you simply wrap elements in an `<ng-template>` without applying a structural directive those elements will simply disappear.

Reviewer 3: ```suggestion
  Note that Angular's `<ng-template>` element defines a template that doesn't render anything by default, if you just wrap elements in an `<ng-template>` without applying a structural directive those elements will not be rendered.
```

  For more information, see the [ng-template API](api/core/ng-template) documentation.

</div>

{@a one-per-element}

> ## One structural directive per element

Author: This section was removed in PR #40015 (which by the way is the PR that made this guide non-generic)

I am not a big fan of moving changes back and forth but I believe the way this was refactored really just lacks clarity and makes things much more difficult to understand.

I put back the section with one or two minor tweaks as I thought it was already well written, if we want I can rewrite this with my own words too.

> Someday you'll want to repeat a block of HTML but only when a particular condition is true.
> You'll _try_ to put both an `*ngFor` and an `*ngIf` on the same host element.
> Angular won't let you. You may apply only one _structural_ directive to an element.
> 
> The reason is simplicity. Structural directives can do complex things with the host element and its descendants.
> When two directives lay claim to the same host element, which one should take precedence?

Reviewer 3: ```suggestion
It's a quite common use-case to repeat a block of HTML but only when a particular condition is true. An intuitive way to do that is to put both an `*ngFor` and an `*ngIf` on the same element. However, since both `*ngFor` and an `*ngIf` are structural directives, this would be treated as an error by the compiler. You may apply only one _structural_ directive to an element.

The reason is simplicity. Structural directives can do complex things with the host element and its descendants.
When two directives lay claim to the same host element, which one should take precedence?
```

named `$implicit`.

Angular sets `let-hero` to the value of the context's `$implicit` property, which `NgFor` has initialized with the hero for the current iteration.

For more information, see the [NgFor API](api/common/NgForOf "API: NgFor") and [NgForOf API](api/common/NgForOf) documentation.

### Creating template fragments with `<ng-template>`

Angular's `<ng-template>` element defines a template that doesn't render anything by default.
With `<ng-template>`, you can render the content manually for full control over how the content displays.

If there is no structural directive and you wrap some elements in an `<ng-template>`, those elements disappear.
In the following example, Angular does not render the middle "Hip!" in the phrase "Hip! Hip! Hooray!" because of the surrounding `<ng-template>`.

<code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (template-tag)" region="template-tag"></code-example>

<div class="lightbox">
  <img src='generated/images/guide/structural-directives/template-rendering.png' alt="template tag rendering">
</div>

> 

Author: I don't think this should sit in the structural directives guide, I changed this (see line 70 above) to a quick alert redirecting to the ng-template api docs (in which I also added a shortened alert with the example shown here)

For example, consider the following structural directive that takes the result of a template expression as an input:

<code-example language="ts" header="IfLoadedDirective">

> export type Loaded<T> = { type: 'loaded', data: T };

Author: I updated this to use `&lt;` and `&gt;` since without them the type annotations are actually not displayed:
![Screenshot at 2022-01-29 21-34-55](https://user-images.githubusercontent.com/61631103/151678285-2c0d7a27-cd51-4e45-a1fa-937af888cae5.png)

(PS: I also split the lines to help in readability)
